pkg/occlient/templates.go

```diff
@@ -0,0 +1,289 @@
+
+func generateGitDeploymentConfig(commonObjectMeta metav1.ObjectMeta, image string, containerPorts []corev1.ContainerPort) appsv1.DeploymentConfig {
+	return appsv1.DeploymentConfig{
+		ObjectMeta: commonObjectMeta,
+		Spec: appsv1.DeploymentConfigSpec{
+			Replicas: 1,
+			Selector: map[string]string{
+				"deploymentconfig": commonObjectMeta.Name,
+			},
+			Template: &corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"deploymentconfig": commonObjectMeta.Name,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Image: image,
+							Name:  commonObjectMeta.Name,
+							Ports: containerPorts,
+						},
+					},
+				},
+			},
+			Triggers: []appsv1.DeploymentTriggerPolicy{
+				{
+					Type: "ConfigChange",
+				},
+				{
+					Type: "ImageChange",
+					ImageChangeParams: &appsv1.DeploymentTriggerImageChangeParams{
+						Automatic: true,
+						ContainerNames: []string{
+							commonObjectMeta.Name,
+						},
+						From: corev1.ObjectReference{
+							Kind: "ImageStreamTag",
+							Name: image,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// generateBuildConfig creates a BuildConfig for Git URL's being passed into Odo
+func generateBuildConfig(commonObjectMeta metav1.ObjectMeta, gitURL string, imageName string, imageNamespace string) buildv1.BuildConfig {
+
+	buildSource := buildv1.BuildSource{
+		Git: &buildv1.GitBuildSource{
+			URI: gitURL,
+		},
+		Type: buildv1.BuildSourceGit,
+	}
+
+	return buildv1.BuildConfig{
+		ObjectMeta: commonObjectMeta,
+		Spec: buildv1.BuildConfigSpec{
+			CommonSpec: buildv1.CommonSpec{
+				Output: buildv1.BuildOutput{
+					To: &corev1.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: commonObjectMeta.Name + ":latest",
+					},
+				},
+				Source: buildSource,
+				Strategy: buildv1.BuildStrategy{
+					SourceStrategy: &buildv1.SourceBuildStrategy{
+						From: corev1.ObjectReference{
+							Kind:      "ImageStreamTag",
+							Name:      imageName,
+							Namespace: imageNamespace,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+//
+// Below is related to SUPERVISORD
+//
+
+// AddBootstrapInitContainer adds the bootstrap init container to the deployment config
+// dc is the deployment config to be updated
+// dcName is the name of the deployment config
+func addBootstrapVolumeCopyInitContainer(dc *appsv1.DeploymentConfig, dcName string) {
+	dc.Spec.Template.Spec.InitContainers = append(dc.Spec.Template.Spec.InitContainers,
+		corev1.Container{
+			Name: "copy-files-to-volume",
+			// Using custom image from bootstrapperImage variable for the initial initContainer
+			Image: dc.Spec.Template.Spec.Containers[0].Image,
+			Command: []string{
+				"sh",
+				"-c"},
+			// Script required to copy over file information from /opt/app-root
+			// Source https://github.com/jupyter-on-openshift/jupyter-notebooks/blob/master/minimal-notebook/setup-volume.sh
+			Args: []string{`
+SRC=/opt/app-root
```

Comment:
Can we move this script to a separate file instead of hard coding here for better maintainability?
Moreover, can the script be copied over instead of being sent as string?

Reply:
nope :-( it has to be there we can't put it to the image because the image is a standard s2i image that we don't want to modify.
Maybe in the future, we will figure out how to solve this in the better way, but currently, this is the best we could come up with